File: csharp/demo/demo15b-deadlock.cs
/*
 * DEADLOCK
 * Version B
 */
using System;
using System.Threading;



class Demo15B : IRunnable
{
    private object resourceA = "resourceA";
    private object resourceB = "resourceB";


    public void run()
    {
        var thFoo = new Thread(() =>
        {
            lock (resourceA)
            {
                Console.WriteLine("foo acquired resource A");
                Thread.Sleep(1000);

                lock (resourceB)
                {
                    Console.WriteLine("foo acquired resource B");
                }
            }
        });


        var thBar = new Thread(() =>
        {
            lock (resourceB)
            {
                Console.WriteLine("bar acquired resource B");
                Thread.Sleep(1000);

                lock (resourceA)
                {
                    Console.WriteLine("bar acquired resource A");
                }
            }
        });


        thFoo.Start();
        thBar.Start();
        thFoo.Join();
        thBar.Join();


        Console.WriteLine("You will never see this statement due to deadlock!");
    }
}

File: python/demo15b_deadlock.py
'''
DEADLOCK
Version B
'''

import time
import threading



mutex_a = threading.Lock()
mutex_b = threading.Lock()



def foo():
    with mutex_a:
        print('foo acquired resource A')
        time.sleep(1)
        with mutex_b:
            print('foo acquired resource B')



def bar():
    with mutex_b:
        print('bar acquired resource B')
        time.sleep(1)
        with mutex_a:
            print('bar acquired resource A')



th_foo = threading.Thread(target=foo)
th_bar = threading.Thread(target=bar)

th_foo.start()
th_bar.start()
th_foo.join()
th_bar.join()

print('You will never see this statement due to deadlock!')
